sql/grant_tables/user.go

```diff
@@ -15,325 +15,364 @@
 package grant_tables
 
 import (
-	"fmt"
 	"time"
 
 	"github.com/dolthub/go-mysql-server/sql"
-	"github.com/dolthub/go-mysql-server/sql/expression"
 	"github.com/dolthub/go-mysql-server/sql/in_mem_table"
-
-	"github.com/dolthub/vitess/go/sqltypes"
-)
-
-const userTblName = "user"
-
-var (
-	userPkCols        = []uint16{0, 1}
-	userUserCols      = []uint16{1}
-	errUserPkAssign   = fmt.Errorf("the primary key for the `user` table expects a host and user string")
-	errUserUserAssign = fmt.Errorf("the secondary key for the `user` table expects a user string")
-
-	userTblSchema sql.Schema
 )
 
-// UserPrimaryKey is a key that represents the primary key for the "user" Grant Table.
-type UserPrimaryKey struct {
-	Host string
-	User string
-}
+// User represents a user from the user Grant Table.
+type User struct {
+	User                string
+	Host                string
+	PrivilegeSet        map[PrivilegeType]struct{}
+	Plugin              string
+	Password            string
+	PasswordLastChanged time.Time
+	Locked              bool
+	Attributes          *string
+	//TODO: add the remaining fields
 
-// UserSecondaryKey is a key that represents the secondary key for the "user" Grant Table, which contains only usernames.
-type UserSecondaryKey struct {
-	User string
+	// IsRole is an additional field that states whether the User represents a role or user. In MySQL this must be a
+	// hidden column, therefore it's represented here as an additional field.
+	IsRole bool
 }
 
-var _ in_mem_table.InMemTableDataKey = UserPrimaryKey{}
-var _ in_mem_table.InMemTableDataKey = UserSecondaryKey{}
+var _ in_mem_table.Entry = (*User)(nil)
 
-// AssignValues implements the interface in_mem_table.InMemTableDataKey.
-func (u UserPrimaryKey) AssignValues(vals ...interface{}) (in_mem_table.InMemTableDataKey, error) {
-	if len(vals) != 2 {
-		return u, errUserPkAssign
+// NewFromRow implements the interface in_mem_table.Entry.
+func (u *User) NewFromRow(ctx *sql.Context, row sql.Row) (in_mem_table.Entry, error) {
+	if err := userTblSchema.CheckRow(row); err != nil {
+		return nil, err
 	}
-	host, ok := vals[0].(string)
-	if !ok {
-		return u, errUserPkAssign
+	//TODO: once the remaining fields are added, fill those in as well
+	var attributes *string
+	passwordLastChanged := time.Now().UTC()
+	if val, ok := row[userTblColIdxMap["User_attributes"]].(string); ok {
+		attributes = &val
 	}
-	user, ok := vals[1].(string)
-	if !ok {
-		return u, errUserPkAssign
+	if val, ok := row[userTblColIdxMap["password_last_changed"]].(time.Time); ok {
+		passwordLastChanged = val
 	}
-	return UserPrimaryKey{
-		Host: host,
-		User: user,
+	return &User{
+		User:                row[userTblColIdxMap["User"]].(string),
```

Comment:
I did it this way so that if the column ordering ever changed in the schema then we wouldn't have to update anything as it would all "just work", but I can make it all constant.